<a href="https://colab.research.google.com/github/AakarshPatel/pytorch-heart-disease/blob/main/01_data_exploration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

In [15]:
data = pd.read_csv('/content/drive/MyDrive/heart_disease/processed.cleveland.data',header=None) # Load data using pandas.

In [32]:
data.shape

(303, 14)

In [16]:
data.columns # Check to see what columns the data has.

Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13], dtype='int64')

In [17]:
data.head() # Get an idea of how the data looks.

0    1    2      3      4    5    6      7    8    9    10   11   12  13
0  63.0  1.0  1.0  145.0  233.0  1.0  2.0  150.0  0.0  2.3  3.0  0.0  6.0   0
1  67.0  1.0  4.0  160.0  286.0  0.0  2.0  108.0  1.0  1.5  2.0  3.0  3.0   2
2  67.0  1.0  4.0  120.0  229.0  0.0  2.0  129.0  1.0  2.6  2.0  2.0  7.0   1
3  37.0  1.0  3.0  130.0  250.0  0.0  0.0  187.0  0.0  3.5  3.0  0.0  3.0   0
4  41.0  0.0  2.0  130.0  204.0  0.0  2.0  172.0  0.0  1.4  1.0  0.0  3.0   0

In [21]:
data.rename(columns={0:'age',1:'sex',2:'cp',3:'trestbps',4:'chol',5:'fbs',
                     6:'restecg',7:'thalach',8: 'exang',9:'oldpeak',10:'slope',
                     11:'ca',12:'thal', 13:'num'},inplace=True) # Rename numerical column headers to more descriptive feature names for the heart disease dataset.

In [22]:
data.head() # Check to see if column names have been updated.

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  
0    3.0  0.0  6.0    0  
1    2.0  3.0  3.0    2  
2    2.0  2.0  7.0    1  
3    3.0  0.0  3.0    0  
4    1.0  0.0  3.0    0

In [23]:
data.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'],
      dtype='object')

In [53]:
column_list = data.columns.tolist() # Convert column names to a list.

['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num']


In [26]:
data.info() # Check to see null and Dyteps in data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        303 non-null    object 
 12  thal      303 non-null    object 
 13  num       303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


In [27]:
# Check to see why columns 11 and 12 have object Dytpes
print("Unique values in 'ca' column:", data['ca'].unique())
print("Unique values in 'thal' column:", data['thal'].unique())

# NEED TO HANDLE MISSING DATA IN COLUMNS 11 AND 12

Unique values in 'ca' column: ['0.0' '3.0' '2.0' '1.0' '?']
Unique values in 'thal' column: ['6.0' '3.0' '7.0' '?']


In [38]:
data.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64

In [30]:
data.describe() # data.describe wont display collumns 11(ca) and 12(12)

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope         num  
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000  
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.937294  
std      0.994971   22.875003    0.469794    1.161075    0.616226    1.228536  
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000  
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000  
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000  
75%      2.000000  166.000000    1.000000    1.600000    2.000000    2.000000  
max      2.000000  202.000000    1.000000    6.200000    3.000000    4.000000

### Initial Observations from Summary Statistics
- Age ranges from 29 to 77 years, with most patients in middle age (mean ~54).
- **Sex:** Skewed toward male patients (~68% male).
- **Chest pain type (`cp`):** Values 1–4 indicate different types of pain; majority are type 3
- **Resting blood pressure (`trestbps`):** 94–200 mm Hg, mean ~132 mm Hg. A few high values may be potential outliers.
-**Cholesterol (`chol`):** 126–564 mg/dl, mean ~247 mg/dl. Some very high readings suggest potential outliers.  
- **Fasting blood sugar (`fbs`):** ~15% of patients have high fasting blood sugar (>120 mg/dl).  
- **Resting ECG (`restecg`):** Most values are 0 or 1, representing normal vs minor abnormalities.  
- **Maximum heart rate achieved (`thalach`):** 71–202 bpm, mean ~150 bpm.  
- **Exercise-induced angina (`exang`):** ~33% of patients show angina during exercise.  
- **ST depression (`oldpeak`):** 0–6.2; higher values indicate greater exercise-induced depression, possible outliers.  
- **Slope of peak ST segment (`slope`):** 1–3, showing upsloping, flat, or downsloping; mostly 1 or 2.  
- **Number of major vessels colored (`ca`):** Contains `?` values; currently treated as object type and will be cleaned later.  
- **Thalassemia (`thal`):** Contains `?` values; to be handled in data cleaning.  
- **Target (`num`):** Min = 0, max = 4, mean ~0.94. Will be binarized for heart disease presence vs absence. Slight class imbalance exists (more patients without disease).

**Notes:**  
- Columns `ca` and `thal` contain missing entries represented as `?`.  
- Outliers in `chol`, `trestbps`, and `oldpeak` should be noted for later analysis.  
- Using only the 14 relevant features aligns with ML research on the Cleveland dataset.



In [49]:
count1 = 0
count2 = 0
count3 = 0
count4 = 0
for index, row_data in data.iterrows(): # Iterate over rows
  value = row_data['num'] # Access 'num' column from the current row Series
  if value == 1:
    count1 += 1
  elif value == 2:
    count2 += 1
  elif value == 3:
    count3 += 1
  else:
    count4 += 1
print(f"num people with presence 1: {count1} \nnum people with presence 2: {count2} \nnum people with presence 3: {count3} \nnum people with presence 4: {count4}")

num people with presence 1: 55 
num people with presence 2: 36 
num people with presence 3: 35 
num people with presence 4: 177


### Target Variable (`num`) Observations

The `num` column represents heart disease diagnosis with values 0–4:

- 0 = no disease  
- 1–4 = disease presence  

Counts of patients:

- Presence 1: 55  
- Presence 2: 36  
- Presence 3: 35  
- Presence 4: 177  

**Insights:**

- There is a slight imbalance in disease presence, with a majority of patients in presence 4.  
- For modeling, this column will be binarized: 0 = no disease, 1–4 = disease presence.

In [57]:
for column in column_list:
  print(f"Unique values in '{column}' column:\n\n", data[column].unique())

Unique values in 'age' column:

 [63. 67. 37. 41. 56. 62. 57. 53. 44. 52. 48. 54. 49. 64. 58. 60. 50. 66.
 43. 40. 69. 59. 42. 55. 61. 65. 71. 51. 46. 45. 39. 68. 47. 34. 35. 29.
 70. 77. 38. 74. 76.]
Unique values in 'sex' column:

 [1. 0.]
Unique values in 'cp' column:

 [1. 4. 3. 2.]
Unique values in 'trestbps' column:

 [145. 160. 120. 130. 140. 172. 150. 110. 132. 117. 135. 112. 105. 124.
 125. 142. 128. 170. 155. 104. 180. 138. 108. 134. 122. 115. 118. 100.
 200.  94. 165. 102. 152. 101. 126. 174. 148. 178. 158. 192. 129. 144.
 123. 136. 146. 106. 156. 154. 114. 164.]
Unique values in 'chol' column:

 [233. 286. 229. 250. 204. 236. 268. 354. 254. 203. 192. 294. 256. 263.
 199. 168. 239. 275. 266. 211. 283. 284. 224. 206. 219. 340. 226. 247.
 167. 230. 335. 234. 177. 276. 353. 243. 225. 302. 212. 330. 175. 417.
 197. 198. 290. 253. 172. 273. 213. 305. 216. 304. 188. 282. 185. 232.
 326. 231. 269. 267. 248. 360. 258. 308. 245. 270. 208. 264. 321. 274.
 325. 235. 257. 164. 141. 252.

**Notes:**
- Several categorical variables are numerically encoded and will need to be treated appropriately during preprocessing.
- Missing values represented by `'?'` appear only in `ca` and `thal` and will be addressed in the data cleaning notebook.